### PR TITLE
Hide beatmap set covers and preview button on explicit beatmaps

### PR DESCRIPTION
**Pull request by frenzibyte**
- [ ] Depends on #11583

Third step in fulfilling #11460 

https://user-images.githubusercontent.com/22781491/105618579-0a3e6300-5dfa-11eb-8ed5-3c20773607ab.mov

Worthy to note:
 - Beatmap set overlay will be affected by this change as well, an upcoming PR will make the overlay only enterable if the user chose to allow explicit content anyways (via popup dialog), so not worth changing anything there at the moment.
 - Currently replicating osu!web's behaviour on this 1:1, not sure about disabling/hiding the download button and how it would feel UX-wise so I decided not to touch it for now.

### CODE REVIEW

**Comment by bdach** [comment]
abstaining from review until the prerequisite is fixed.

File: osu.Game/Online/ConfigExplicitContentPermission.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+using osu.Framework.Bindables;
+using osu.Game.Configuration;
+
+namespace osu.Game.Online
+{
+    /// <summary>
+    /// A type of <see cref="IExplicitContentPermission"/> that binds <see cref="UserAllowed"/> to <see cref="OsuSetting.ShowOnlineExplicitContent"/>'s value directly.
+    /// </summary>
+    internal class ConfigExplicitContentPermission : IExplicitContentPermission
```
Comment (peppy) [inline osu.Game/Online/ConfigExplicitContentPermission.cs:12]: I don't get this whole direction. Why is this required as opposed to binding to the config bindable directly?

Comment (frenzibyte) [reply osu.Game/Online/ConfigExplicitContentPermission.cs:12]: It appears I haven't explained the direction and the idea behind `IExplicitContentPermission` in the PR description, so will do here:

`IExplicitContentPermission` is an interface that should be DI'd by components to know whether they should display / play explicit content or not, based on the bindable `UserAllowed`.

`DialogExplicitContentPermission` (which will be in another PR) is a class that can be cached at an overlay / screen level for popping up a dialog to the user asking for explicit content permission, in a similar fashion to osu!web's behaviour:
image

And will manage the `UserAllowed` bindable based on the chosen option.

`ConfigExplicitContentPermission` is a class cached at a game level for components which are not covered by a `DialogExplicitContentPermission` or otherwise (e.g. osu!direct panels), in which the `UserAllowed` will be solely based on the value of the config setting `OsuSetting.ShowOnlineExplicitContent`.

Comment (peppy) [reply osu.Game/Online/ConfigExplicitContentPermission.cs:12]: And there are going to be other implementations of `IExplicitContentPermission`..? If not I'd remove the interface.

Also the xmldoc for this reads really weirdly. It shouldn't say what it's linking to what, but instead what the overall purpose is. Something like "A component which exposes as a bindable whether the current user has allowed display of explicit content".

Comment (frenzibyte) [reply osu.Game/Online/ConfigExplicitContentPermission.cs:12]: Not really. The interface exists because there are two classes, not one which can be used solely.

Comment (peppy) [reply osu.Game/Online/ConfigExplicitContentPermission.cs:12]: What's the second class? I'm not sure why there is two levels of permission in the first place, does this requirement come from some specific use case?

Comment (frenzibyte) [reply osu.Game/Online/ConfigExplicitContentPermission.cs:12]: I've explained it [above](https://github.com/ppy/osu/pull/11585#discussion_r644505418), the classes are `DialogExplicitContentPermission` and `ConfigExplicitContentPermission`, the former pops up a dialog and decides the `UserAllowed` value based on the selected option, and the latter binds `UserAllowed` to the config value directly.

There aren't meant to be two levels of permissions exactly, this is done so that components which are not covered by a `DialogExplicitContentPermission` will be covered by `ConfigExplicitContentPermission` instead, that's why it exists in the first place.

Things may get messy for each component if `ConfigExplicitContentPermission` doesn't exist, as each one will have to also resolve the config manager and bind to the explicit setting as a secondary check, that's why I thought about adding a `ConfigExplicitContentPermission` and caching it at a game-wide level, which makes sense for me.

Comment (bdach) [reply osu.Game/Online/ConfigExplicitContentPermission.cs:12]: I'm not personally convinced by any of the above that the "dialog permission" needs to be a thing. If it is to always be used in a context in which explicit content off means that the entire overlay content is hidden, then that can be achieved using a single drawable/container that shows the dialog if explicit content is off, and swaps the dialog out for the actual content if explicit content is enabled either via settings or via the dialog shown.

Comment (frenzibyte) [reply osu.Game/Online/ConfigExplicitContentPermission.cs:12]: Where do you imagine it to be placed? in each component (`PlayButton` and `BeatmapSetCover`)? or at a central level with some logic to get the components that should be hidden?

Also, there may be components that should handle explicit content differently, rather than just vanishing away from the scene. (such as displaying a custom texture on beatmap set cover, graying out play button, etc.)

Moreover, how do you imagine that working for beatmap listing panels of explicit beatmaps? where a dialog should not be requested for seeing them, and instead should be displayed based on the config setting only.

All of these, and importantly the last one, are why I have chosen the `IExplicitContentPermission` path, so that the actual components will be controlled by a cached `IExplicitContentPermission` implementation by only keeping their eyes on the bindable exposed by it, and change their state whenever the bindable changes.

Comment (bdach) [reply osu.Game/Online/ConfigExplicitContentPermission.cs:12]: I was only talking about the dialog permission case. If the dialog implementation is eliminated, then only one implementation of that interface remains, and the interface's existence also becomes questionable.

For what it's worth even if you implemented `DialogExplicitContentPermission`, then you'd still need a visual component that does the toggling in the same exact way I said. At that point splitting them into two seems like overengineering.

Comment (bdach) [reply osu.Game/Online/ConfigExplicitContentPermission.cs:12]: In other words: I believe that you only ever really should need
```csharp
config.GetBindable<bool>(OsuSetting.ShowOnlineExplicitContent)
```
unless you have (and can describe succinctly) a solid case that proves otherwise. I guess you can wrap that one bindable in a cached component if you insist, but that doesn't really change the fact that this tiered permission setup does not really seem necessary, in the context of things as they are presented currently.

Comment (frenzibyte) [reply osu.Game/Online/ConfigExplicitContentPermission.cs:12]: There is a case against that, which is temporarily allowing the user to view the content once, that will require more than just a bindable, and is literally why I have all this `IExplicitContentPermission` setup here.

Comment (bdach) [reply osu.Game/Online/ConfigExplicitContentPermission.cs:12]: Details, please. Are you still talking about the overlay case? The "show this beatmap once" case? Does a drawable component as I described above not solve that exact issue?

Comment (bdach) [reply osu.Game/Online/ConfigExplicitContentPermission.cs:12]: Okay, I think I get it, but it would be nice to not get generic responses to my questions or not have to come up with the answer on my own. The answer to that is that in my proposal, nested components of the overlay (such as beatmap covers, etc.) would be reading the config bindable locally. So the override needs to work at a screen/overlay-level, because otherwise you'd show the contents of the overlay but the child components would know nothing about the override.

Comment (frenzibyte) [reply osu.Game/Online/ConfigExplicitContentPermission.cs:12]: Well, I've answered to [your proposal](https://github.com/ppy/osu/pull/11585#discussion_r644669282) with https://github.com/ppy/osu/pull/11585#discussion_r644687881, and I've answered to [your question](https://github.com/ppy/osu/pull/11585#discussion_r644693040) with https://github.com/ppy/osu/pull/11585#discussion_r644696725.

I guess it's the last response that's generic https://github.com/ppy/osu/pull/11585#discussion_r644696725, I thought that's pretty simplified way to explain it, it just needs more than a config bindable for the component to update itself, because of the temporary permission case, and that's why the whole `IExplicitContentPermission` setup is here, rather than just binding to a config setting.

**Comment by smoogipoo** [comment]
Going to close due to inactivity + going to require a herculean effort to merge due to the panels having since been rewritten. Best start off fresh.
